lib/openssh/sshd.go

```diff
@@ -0,0 +1,243 @@
+)
+
+const DefualtRestartCommand = "systemctl restart sshd"
+
+const (
+	// TeleportKey is the name the OpenSSH private key
+	TeleportKey = "ssh_host_teleport_key"
+	// TeleportKey is the name the OpenSSH cert
+	TeleportCert = "ssh-host-teleport-cert.pub"
+	// TeleportOpenSSHCA is the path to the Teleport OpenSSHCA
+	TeleportOpenSSHCA = "teleport_openssh_ca.pub"
+)
+
+type sshdBackendOperations interface {
+	restart() error
+	checkConfig(path string) error
+}
+
+// SSHD is used to update the OpenSSH config
+type SSHD struct {
+	sshd sshdBackendOperations
+}
+
+// NewSSHD initializes SSHD
+func NewSSHD(restartCmd string) SSHD {
+	return SSHD{
+		sshd: &sshdBackend{
+			restartCmd: restartCmd,
+		},
+	}
+}
+
+// SSHDConfigUpdate is the list of options to be set in the Teleport OpenSSH config
+type SSHDConfigUpdate struct {
+	// SSHDConfigPath is the path to the OpenSSH sshd_config file
+	SSHDConfigPath string
+	// TeleportSSHDConfigPath is the path where Teleport will write its own sshd config options
+	TeleportSSHDConfigPath string
+	// OpenSSHCAPath is the path to which Teleport OpenSSHCA will be written
+	OpenSSHCAPath string
+	// HostKeyPath is the path to the Teleport Host Key
+	HostKeyPath string
+	// HostCertPath is the path to the Teleport OpenSSH cert
+	HostCertPath string
+}
+
+var sshdConfigTmpl = template.Must(template.New("").Parse(`# Created by 'teleport join openssh', do not edit
+TrustedUserCAKeys {{ .OpenSSHCAPath }}
+HostKey {{ .HostKeyPath }}
+HostCertificate {{ .HostCertPath }}
+`))
+
+func fmtSSHDConfigUpdate(u SSHDConfigUpdate) (string, error) {
+	buf := &bytes.Buffer{}
+	if err := sshdConfigTmpl.Execute(buf, u); err != nil {
+		return "", trace.Wrap(err)
+	}
+	return buf.String(), nil
+}
+
+// UpdateConfig updates the sshd_config file if needed and writes the
+// teleport specific configuration
+func (s *SSHD) UpdateConfig(u SSHDConfigUpdate, restart bool) error {
+	configUpdate, err := fmtSSHDConfigUpdate(u)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if err := writeTempAndRename(u.TeleportSSHDConfigPath, s.sshd.checkConfig, []byte(configUpdate)); err != nil {
+		return trace.Wrap(err)
+	}
+
+	needsUpdate, err := checkSSHDConfigAlreadyUpdated(u.SSHDConfigPath, sshdConfigInclude)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	if needsUpdate {
+		if err := prependToSSHDConfig(u.SSHDConfigPath, sshdConfigInclude); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	if restart {
+		if err := s.sshd.restart(); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
+}
+
+// WriteKeys writes the OpenSSH keys and CA from the Identity and the
+// OpenSSH CA to disk for the OpenSSH daemon to use
+func WriteKeys(keysdir string, id *auth.Identity, cas []types.CertAuthority) error {
+	if err := os.MkdirAll(keysdir, 0o700); err != nil {
+		return trace.ConvertSystemError(err)
+	}
+	if err := os.WriteFile(filepath.Join(keysdir, TeleportKey), id.KeyBytes, 0600); err != nil {
```

Comment:
I feel like all of these should use `writeTempAndRename` just so we don't make a machine unconnectable if the power is cut at exactly the wrong time.